File: sudoku.rb
require 'colorize'

class Sudoku
  def initialize 
    @board = Hash.new
    @known_keys = []
    @sub_range = [[0,1,2], [3,4,5], [6,7,8]]
    File.open('sample') do |f|
      row = 0
      while line = f.gets
        line.split.each_with_index do |number, col|
          value = number.to_i
          @board[[col, row]] = value
          @known_keys << [col, row] if value > 0
          #print "#{value} "
        end
        #puts
        row += 1
      end
    end
    show_result
    puts
    puts "Just a moment..."
    puts
    #puts @known_keys.to_s
  end

  def check_row(x, y)
    row = (0..8).to_a.product([y]).collect{|key| @board[key]}
    #print row
    row.delete 0
    return row == row.uniq
  end

  def check_column(x, y)
    col = [x].product((0..8).to_a).collect{|key| @board[key]}
    col.delete 0
    return col == col.uniq
  end

  def check_sub(x, y)
    x_r = @sub_range[ x / 3 ]
    y_r = @sub_range[ y / 3 ]
    nine = x_r.product(y_r).collect{|key| @board[key] }
    nine.delete 0
    return nine == nine.uniq
  end

  def try(step=0) 
    #print "\n\n"
    #puts step

    if step == 81
      show_result
      return true
    end

    y = step / 9
    x = step % 9
    key = [x, y]
    return try( step + 1 ) if @known_keys.include? key

    while true
      @board[key] += 1
      if @board[key] == 10
        @board[key] = 0
        return false
      end
      return true if check_row(x,y) && check_column(x,y) && check_sub(x,y) && try(step + 1)
      
    end    
  end
  
  def show_result()
    #(0..8).to_a.product((0..8).to_a).collect{|key| @board[key]}.each
    for y in (0..8)
      for x in (0..8)
        if @known_keys.include? [x,y]
          print @board[[x,y]].to_s.red
        else
          print @board[[x,y]] > 0 ? @board[[x,y]] : '.'
        end
        print ' '
      end
      print "\n"
    end
  end
end

s = Sudoku.new
s.check_row 1,3
s.try

File: sudoku.py
import sys

def non_zero(x):
  return x > 0 and x

class Sudoku:
  def __init__(self, puzzle):
    self.board = {}
    self.known_keys = []
    self.sub_range = [[0,1,2], [3,4,5], [6,7,8]]
    self.all = range(1, 10)
    self.all_index = range(0, 9)
    with open(puzzle) as f:
      for row, line in enumerate( f.readlines() ):
        for col, number in enumerate( line.split() ):
          value = int(number)
          self.board[ (col, row) ] = value
          if value > 0:
            self.known_keys.append( (col, row) )
    self.show_results()


  def show_results(self):
    for row in range(0, 9):
      for col in range(0, 9):
        print( self.board[ (col, row) ] > 0 and self.board[ (col, row) ] or '.', end=' ' )
      print()

  def solve(self):
    print()
    print("Just a moment...")
    print()
    self.try_next()



  def get_row(self, r):
    row = [ self.board[ (x, y) ] for x in self.all_index for y in [r] ]
    return filter( non_zero, row )

  def get_col(self, c):
    col = [ self.board[ (x, y) ] for x in [c] for y in self.all_index ]
    return filter( non_zero, col)

  def get_nine(self, c, r):
    nine = [ self.board[ (x, y) ] for x in self.sub_range[ int(c / 3) ]  for y in self.sub_range[ int(r / 3) ] ]
    return filter( non_zero, nine)

  def try_next(self, step=0):
    if step == 81:
      self.show_results()
      return True
    x = int(step / 9)
    y = step % 9
    key = (x, y)

    if key in self.known_keys:
      return self.try_next( step + 1)

    guesses = set(self.all) - set( self.get_row(y) ) - set( self.get_col(x) ) - set( self.get_nine(x,y) )
    for g in guesses:
      self.board[ key ] = g
      if self.try_next( step + 1):
        return True
    self.board[ key ] = 0
    return False


if __name__ == "__main__":
  s = Sudoku(sys.argv[1])
  s.solve()
